extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java

```diff
@@ -0,0 +1,138 @@
+import com.google.protobuf.util.JsonFormat;
+import io.druid.data.input.ByteBufferInputRowParser;
+import io.druid.data.input.InputRow;
+import io.druid.data.input.MapBasedInputRow;
+import io.druid.data.input.impl.ParseSpec;
+import io.druid.java.util.common.parsers.ParseException;
+import io.druid.java.util.common.parsers.Parser;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+@JsonTypeName("protobuf")
+public class ProtoBufInputRowParser implements ByteBufferInputRowParser {
+  //private static final Logger log = new Logger(ProtoBufInputRowParser.class);
+
+  private final ParseSpec parseSpec;
+  private Parser<String, Object> parser;
+  private final String descriptorFilePath;
+  private final String protoMessageType;
+  private Descriptor descriptor;
+
+
+  @JsonCreator
+  public ProtoBufInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("descriptor") String descriptorFilePath,
+      @JsonProperty("protoMessageType") String protoMessageType
+  ) {
+    this.parseSpec = parseSpec;
+    this.descriptorFilePath = descriptorFilePath;
+    this.protoMessageType = protoMessageType;
+    this.parser = parseSpec.makeParser();
+    this.descriptor = getDescriptor(descriptorFilePath);
+  }
+
+  @Override
+  public ParseSpec getParseSpec() {
+    return parseSpec;
+  }
+
+  @Override
+  public ProtoBufInputRowParser withParseSpec(ParseSpec parseSpec) {
+    return new ProtoBufInputRowParser(parseSpec, descriptorFilePath, protoMessageType);
+  }
+
+  @Override
+  public InputRow parse(ByteBuffer input) {
+    String json;
+    try {
+      DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
+      json = JsonFormat.printer().print(message);
+      // log.trace(json);
+    } catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+
+    Map<String, Object> record = parser.parse(json);
+    return new MapBasedInputRow(
+        parseSpec.getTimestampSpec().extractTimestamp(record),
+        parseSpec.getDimensionsSpec().getDimensionNames(),
+        record
+    );
+  }
+
+  private Descriptor getDescriptor(String descriptorFilePath) {
+    InputStream fin;
+
+    fin = this.getClass().getClassLoader().getResourceAsStream(descriptorFilePath);
+    if (fin == null) {
+      URL url = null;
+      try {
+        url = new URL(descriptorFilePath);
+      } catch (MalformedURLException e) {
+        throw new ParseException(e, "descriptor has to be in the classpath or URL:" + descriptorFilePath);
+      }
+      try {
+        fin = url.openConnection().getInputStream();
+      } catch (IOException e) {
+        throw new ParseException(e, "Cannot open descriptor file: " + url.toString());
+      }
+    }
+
+    DynamicSchema dynamicSchema = null;
+    try {
```

Comment:
The old code did `DescriptorProtos.FileDescriptorSet.parseFrom(fin)` and this new code instead uses DynamicSchema from a new protobuf-dynamic library. What's the rationale for this change? Is it compatible with descriptor files people might have been using with the old code?